zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnJobBehavior.java

```diff
@@ -159,9 +166,98 @@ private Either<Failure, List<LinkedResourceProps>> evalLinkedResourceProps(
   }
 
   private String resolveLinkedResourceKey(
-      final LinkedResource linkedResource, final long scopeKey) {
-    // TODO Add implementation in the next PR
-    return "";
+      final LinkedResource linkedResource, final BpmnElementContext context, final long scopeKey) {
+    return findLinkedResource(
+            linkedResource.getResourceId(),
+            linkedResource.getBindingType(),
+            linkedResource.getVersionTag(),
+            context,
+            scopeKey)
+        .map(PersistedResource::getResourceKey)
+        .get()
+        .toString();
+  }
+
+  private Either<Failure, PersistedResource> findLinkedResource(
+      final String resourceId,
+      final ZeebeBindingType bindingType,
+      final String versionTag,
+      final BpmnElementContext context,
+      final long scopeKey) {
+    return switch (bindingType) {
+      case deployment -> findResourceByIdInSameDeployment(resourceId, context, scopeKey);
+      case latest -> findLatestResourceById(resourceId, context.getTenantId(), scopeKey);
+      case versionTag ->
+          findResourceByIdAndVersionTag(resourceId, versionTag, context.getTenantId(), scopeKey);
+    };
+  }
+
+  private Either<Failure, PersistedResource> findResourceByIdInSameDeployment(
+      final String resourceId, final BpmnElementContext context, final long scopeKey) {
+    return stateBehavior
+        .getDeploymentKey(context.getProcessDefinitionKey(), context.getTenantId())
+        .flatMap(
+            deploymentKey ->
+                resourceState
+                    .findResourceByIdAndDeploymentKey(
+                        resourceId, deploymentKey, context.getTenantId())
+                    .<Either<Failure, PersistedResource>>map(Either::right)
+                    .orElseGet(
+                        () ->
+                            Either.left(
+                                new Failure(
+                                    String.format(
+                                        """
+                                        Expected to use a resource with id '%s' with binding type 'deployment', \
+                                        but no such resource found in the deployment with key %s which contained the current process. \
+                                        To resolve this incident, migrate the process instance to a process definition \
+                                        that is deployed together with the intended resource to use.\
```

Comment:
🔧 We can be more explicit about what was attempted

```suggestion
                                        Expected to link a resource with id '%s' and binding type 'deployment', \
                                        but no such resource found in the deployment with key %s which contained the current process. \
                                        To resolve this incident, migrate the process instance to a process definition \
                                        that is deployed together with the intended resource to use.\
```